### PR TITLE
Examples refactor

### PR DESCRIPTION
**Pull request by ParthSareen**
- Reorganized examples and added a readme
- Added some missing examples

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@ParthSareen thanks for putting things together, i will take a look and share some comments.

File: README.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 The Ollama Python library provides the easiest way to integrate Python 3.8+ projects with [Ollama](https://github.com/ollama/ollama).
 
+## Prerequisites
+
+- [Ollama](https://ollama.com/download) should be installed and running
+- Pull a model to use with the library: `ollama pull <model>` e.g. `ollama pull llama3.1`
```
Comment (mxyng) [inline README.md:8]: llama3.2 should be used for all examples where appropriate

```suggestion
- Pull a model to use with the library: `ollama pull <model>` e.g. `ollama pull llama3.2`
```

File: README.md [hunk 2]
```diff
@@ -2,6 +2,12 @@
...
+  - See [Ollama models](https://ollama.com/models) for more information on the models available.
```
Comment (mxyng) [inline README.md:9]:
```suggestion
  - See [Ollama.com](https://ollama.com/search) for more information on the models available.
```

File: README.md [hunk 3]
```diff
@@ -11,24 +17,32 @@ pip install ollama
 ## Usage
 
 ```python
-import ollama
-response = ollama.chat(model='llama3.1', messages=[
+from ollama import chat
+from ollama._types import ChatResponse
```
Comment (mxyng) [inline README.md:21]: `ollama._types` should not be imported directly. `ChatResponse` is exported explicitly by `ollama`. Also why the change to `from ollama import ...`?

```suggestion
from ollama import chat, ChatResponse
```

File: README.md [hunk 4]
```diff
@@ -38,6 +52,63 @@ for chunk in stream:
...
+- `host`: The Ollama host (default: `http://localhost:11434`)
+- `timeout`: The timeout for requests (default: `None`)
+- `follow_redirects`: Whether to follow redirects (default: `True`)
+- `headers`: Additional headers to send with requests (default: `{}`)
+
+```python
+from ollama import Client
+# default client
+client = Client()
+# or custom client
```
Comment (mxyng) [inline README.md:68]: This section is on custom clients so mentioning the default client seems unnecessary

File: README.md [hunk 5]
```diff
@@ -38,6 +52,63 @@ for chunk in stream:
...
 ```
 
+## Custom client
+
+A custom client can be created with the following fields:
+
+- `host`: The Ollama host (default: `http://localhost:11434`)
+- `timeout`: The timeout for requests (default: `None`)
+- `follow_redirects`: Whether to follow redirects (default: `True`)
+- `headers`: Additional headers to send with requests (default: `{}`)
```
Comment (mxyng) [inline README.md:62]: The user is not restricted to these fields. Any extra kwargs are passed verbatim to `httpx.Client`. These are mentioned only because they're default values differ from `httpx`

File: examples/README.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Running Examples
+
+Run the examples in this directory with:
+```sh
+# Ensure you are in the examples directory
+cd examples/
+
+# Run example
+python3 <example>.py
```
Comment (mxyng) [inline examples/README.md:9]:
```suggestion
# Run example
python3 examples/<example>.py
```

File: examples/list.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+from ollama import list
+from ollama._types import ListResponse
+
+response: ListResponse = list()
+
+for model in response.models:
+  if model.details:
+    print(f'Name: {model.model}')
```
Comment (mxyng) [inline examples/list.py:8]: fstrings are unnecessary here

```suggestion
    print('Name:', model.model)
```

File: examples/list.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+from ollama import list
+from ollama._types import ListResponse
+
+response: ListResponse = list()
+
+for model in response.models:
+  if model.details:
```
Comment (mxyng) [inline examples/list.py:7]: This should only apply to the lines that use `model.details`

File: examples/list.py (hunk 7 above)
Comment (mxyng) [reply examples/list.py:8]: For readability, every line other than `Name` should be indented

File: examples/list.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    print(f'Name: {model.model}')
+    print(f'Size (MB): {(model.size.real / 1024 / 1024):.2f}')
+    print(f'Format: {model.details.format}')
+    print(f'Family: {model.details.family}')
+    print(f'Parameter Size: {model.details.parameter_size}')
+    print(f'Quantization Level: {model.details.quantization_level}')
+    print('-' * 50)
```
Comment (mxyng) [inline examples/list.py:14]: I'm not a fan of this. On narrower terminals, it'll wrap. Better to replace this with a simple newline

File: examples/ps.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  if progress.get('status') in progress_states:
+    continue
+  progress_states.add(progress.get('status'))
+  print(progress.get('status'))
+
+print('\n')
+
+
+response: ProcessResponse = ps()
+for model in response.models:
```
Comment (mxyng) [inline examples/ps.py:17]: similar comment as list example

File: examples/README.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+```
+- [create.py](create.py) 
+
+See [ollama/docs/modelfile.md](https://github.com/ollama/ollama/blob/main/docs/modelfile.md) for more information on the Modelfile format.
+
+
+### Ollama Embed - Generate embeddings with a model
+- [embed.py](embed.py)
+
+### Fill in the middle
```
Comment (mxyng) [inline examples/README.md:60]: This should be grouped with generate/chat

File: examples/tools.py
Comment (mxyng) [inline examples/tools.py:1]: Any sync/async examples should match to highlight only the differences between sync and async. Keep other differences to a minimum

File: examples/README.md (hunk 6 above)
Comment (mxyng) [reply examples/README.md:9]: I'm not sure I like having all the examples in the same directory. Subdirectories are self contained. Examples like pull and create and provide additional information to the user without it bleeding into other examples

File: README.md (hunk 3 above)
Comment (ParthSareen) [reply README.md:21]: Oh you're right let me fix

File: README.md (hunk 5 above)
Comment (ParthSareen) [reply README.md:62]: Okay - let me make that callout instead

File: examples/ps.py (hunk 10 above)
Comment (ParthSareen) [reply examples/ps.py:17]: Not sure what you mean - no details section in this response object

File: examples/tools.py
Comment (ParthSareen) [reply examples/tools.py:1]: Yeah - will make a pass and standardize

File: examples/README.md (hunk 6 above)
Comment (ParthSareen) [reply examples/README.md:9]: I gave this some thought too and tried both things. While you're right individual directories would cause less bleeding into each other - it's just easier to run and play around with examples without having to jump between directories.
I can imagine newer users go into each directory to run the example lol.

File: examples/ps.py (hunk 10 above)
Comment (mxyng) [reply examples/ps.py:17]: I meant similar output format to list example

File: README.md [hunk 12]
```diff
@@ -11,25 +17,33 @@ pip install ollama
...
 ```
 
+See [_types.py](ollama/_types.py) for more information on the response types.
+
 ## Streaming responses
 
-Response streaming can be enabled by setting `stream=True`, modifying function calls to return a Python generator where each part is an object in the stream.
+Response streaming can be enabled by setting `stream=True`
+
+Streaming Tool/Function calling is not yet supported.
```
Comment (mxyng) [inline README.md:40]: You can call out things in GitHub markdown with `> [!TAG]`

```suggestion
> [!NOTE]
> Streaming Tool/Function calling is not yet supported.
```

File: README.md [hunk 13]
```diff
@@ -11,25 +17,33 @@ pip install ollama
...
+# or access fields directly from the response object
+print(response.message.content)
 ```
 
+See [_types.py](ollama/_types.py) for more information on the response types.
+
 ## Streaming responses
 
-Response streaming can be enabled by setting `stream=True`, modifying function calls to return a Python generator where each part is an object in the stream.
+Response streaming can be enabled by setting `stream=True`
```
Comment (mxyng) [inline README.md:38]:
```suggestion
Response streaming can be enabled by setting `stream=True`.
```

File: examples/ps.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+  print(f'Model: {model.model}')
+  print(f'Digest: {model.digest}')
+  print(f'Expires at: {model.expires_at}')
+  print(f'Size: {model.size}')
+  print(f'Size vram: {model.size_vram}')
+  print(f'Details: {model.details}')
+
+  print('---' * 10)
```
Comment (ParthSareen) [inline examples/ps.py:25]: fixing this too

File: README.md [hunk 15]
```diff
@@ -38,20 +52,70 @@ for chunk in stream:
   print(chunk['message']['content'], end='', flush=True)
 ```
 
+## Custom client
+A custom client can be created with the fields from the [httpx client](https://www.python-httpx.org/api/#client).
```
Comment (mxyng) [inline README.md:57]: This reads weird. Focus on a custom Ollama client until `extra arguments are passed to httpx.Client` at the end

File: examples/async-tools.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+  response: ChatResponse = await client.chat(
+    'llama3.1',
+    messages=[{'role': 'user', 'content': prompt}],
+    tools=[add_two_numbers, subtract_two_numbers_tool],
+  )
+
+  available_functions = {
+    'add_two_numbers': add_two_numbers,
+    'subtract_two_numbers': subtract_two_numbers,
+  }
```
Comment (mxyng) [inline examples/async-tools.py:59]:
```suggestion
  available_functions = {
    'add_two_numbers': add_two_numbers,
    'subtract_two_numbers': subtract_two_numbers,
  }

  response: ChatResponse = await client.chat(
    'llama3.1',
    messages=[{'role': 'user', 'content': prompt}],
    tools=available_functions.values(),
  )
```

File: README.md (hunk 15 above)
Comment (ParthSareen) [reply README.md:57]: Yeah reading this again it doesn't read well- will workshop

File: examples/async-tools.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    # There may be multiple tool calls in the response
+    for tool in response.message.tool_calls:
+      # Ensure the function is available, and then call it
+      if tool.function.name in available_functions:
+        print(f'Calling function {tool.function.name}')
+        print(f'Arguments: {tool.function.arguments}')
+        function_to_call = available_functions[tool.function.name]
+        print(f'Function output: {function_to_call(**tool.function.arguments)}')
+      else:
+        print(f'Function {tool.function.name} not found')
```
Comment (mxyng) [inline examples/async-tools.py:71]:
```suggestion
      if function_to_call := available_functions.get(tool.function.name):
        print('Calling function', tool.function.name)
        print('Arguments:', tool.function.arguments)
        print('Function output:', function_to_call(**tool.function.arguments))
      else:
        print(f'Function {tool.function.name} not found')
```

Comment (ParthSareen) [reply examples/async-tools.py:71]: I'd actually avoid using walrus op for examples.

Will do a block like:
```py
function_to_call = available_functions.get(tool.function.name, None)
    if function_to_call:
      print(f'Function output: {function_to_call(**tool.function.arguments)}')
    else:
      print(f'Function {tool.function.name} not found')
```

File: examples/chat-with-history.py [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+while True:
+  user_input = input('Chat with history: ')
+  response = chat(
+    'llama3.2',
+    messages=messages
+    + [
+      {'role': 'user', 'content': user_input},
+    ],
+  )
```
Comment (mxyng) [inline examples/chat-with-history.py:31]: This doesn't add the user input as a message for future calls to chat

```suggestion
  user_input = input('Chat with history: ')
  messages.append({
    'role': 'user',
    'content': user_input,
  })
  response = chat(
    'llama3.2',
    messages=messages,
  )
```

File: README.md [hunk 19]
```diff
@@ -38,20 +53,68 @@ for chunk in stream:
   print(chunk['message']['content'], end='', flush=True)
 ```
 
+## Custom client
+A custom client can be created by instantiating `Client` or `AsyncClient` from `ollama`.
+
+All extra kwargs are passed into the [httpx client](https://www.python-httpx.org/api/#client).
```
Comment (mxyng) [inline README.md:59]: nit:

```suggestion
All extra keyword arguments are passed into the [`httpx.Client`](https://www.python-httpx.org/api/#client).
```

**Review by mxyng — APPROVED** [review]
